### PR TITLE
PTL-1546 criteria matching

### PR DESCRIPTION
**Pull request by patrickoneill-genesis**
Draft for criteria matching. Really not fully certain what we should write for this.

In my opinion, it's just a Groovy expression string generator to create criteria strings to use with Charts, Grids, Entity management and connect snapshot/stream

The module itself is not well documented either

### CODE REVIEW

File: docs/001_develop/03_client-capabilities/011_criteria/index.mdx [hunk 1]
```diff
@@ -1,11 +1,42 @@
...
+id: foundation-criteria
+keywords: [web, criteria, foundation criteria, frontend, ui]
 tags:
-- criteria
-sidebar_position: 11
+  - web
+  - criteria
+  - foundation criteria
+  - frontend
+  - ui
```
Comment (matteematt) [inline docs/001_develop/03_client-capabilities/011_criteria/index.mdx:11]: Can you add more tags and keywords to all these pages (groovy etc.). Can use AI to help I find. We want to have more keywords to aid searchability in docs

File: docs/001_develop/03_client-capabilities/011_criteria/index.mdx [hunk 2]
```diff
@@ -1,11 +1,42 @@
...
+```
+
+## [API Docs](./docs/api/index.md)
+
+## License
+
+Note: this project provides front-end dependencies and uses licensed components listed in the next section; thus, licenses for those components are required during development. Contact [Genesis Global](https://genesis.global/contact-us/) for more details.
+
+### Licensed components
+Genesis low-code platform
```
Comment (matteematt) [inline docs/001_develop/03_client-capabilities/011_criteria/index.mdx:42]: Can get rid of this
```suggestion
```

File: docs/001_develop/03_client-capabilities/011_criteria/01_overview.mdx
Comment (matteematt) [inline docs/001_develop/03_client-capabilities/011_criteria/01_overview.mdx:1]: Can you either turn this page into the index page to match the format of the other documents please? Or if not just get rid of the index page for this module and just link to the API docs from here

File: docs/001_develop/03_client-capabilities/011_criteria/02_criteria-builder.mdx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+keywords: [foundation, ui, comms, criteria, filtering]
+tags:
+  - connected
+  - ui
+  - comms
+  - criteria
+  - filtering
+---
+
+The `CriteriaBuilder` is a utility class designed to build objects used in creating query criteria in a Groovy expression format. These criteria can be used in requests to backend services.
```
Comment (matteematt) [inline docs/001_develop/03_client-capabilities/011_criteria/02_criteria-builder.mdx:14]: Add a link to the API doc for http://localhost:8080/develop/client-capabilities/criteria/docs/api/foundation-criteria.criteriabuilder/ in this page somewhere

File: docs/001_develop/03_client-capabilities/011_criteria/03_expression-buillder.mdx
Comment (matteematt) [inline docs/001_develop/03_client-capabilities/011_criteria/03_expression-buillder.mdx:1]: Add a link to expression builder API docs http://localhost:8080/develop/client-capabilities/criteria/docs/api/foundation-criteria.expressionbuilder/

File: docs/001_develop/03_client-capabilities/011_criteria/02_criteria-builder.mdx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+const tradeIdEqualsExpression = new ExpressionBuilder()
+  .withField('TRADE_ID')
+  .withValue('TR1234')
+  .withSerialiser(Serialisers.EQ)
+  .build()
+```
+:::
+
+
+<table>
```
Comment (matteematt) [inline docs/001_develop/03_client-capabilities/011_criteria/02_criteria-builder.mdx:32]: Screenshot 2024-11-29 at 11 19 22

Nitpicky but there is something wrong with the formatting of the table header, it isn't coloured like everything else.

ALso can you format the code example lines like normal (left indented with space above and below the example) as this allows it to render in other markdown viewers like github too

File: docs/001_develop/03_client-capabilities/011_criteria/03_expression-buillder.mdx [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  - comms
+  - expression
+  - filtering
+---
+
+The `ExpressionBuilder` is a utility class to build criteria objects that are used with the `CriteriaBuilder`
+
+## Methods
+
+<table>
```
Comment (matteematt) [inline docs/001_develop/03_client-capabilities/011_criteria/03_expression-buillder.mdx:18]: Same comment about the table here

File: docs/001_develop/03_client-capabilities/011_criteria/04_form-criteria-builder.mdx [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+keywords: [foundation, ui, comms, criteria, filtering]
+tags:
+  - connected
+  - ui
+  - comms
+  - criteria
+  - filtering
+---
+
+The `Form
```
Comment (matteematt) [inline docs/001_develop/03_client-capabilities/011_criteria/04_form-criteria-builder.mdx:14]: Have you missed checking something in or is this page unneccary?

File: docs/001_develop/03_client-capabilities/011_criteria/05_serialisers.mdx [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+  - comms
+  - criteria
+  - filtering
+  - serialisers
+---
+
+`Serialisers` are a set of utility functions for creating Groovy expressions for different operators e.g. greater han, less than, equals etc
+
+## Methods
+
```
Comment (matteematt) [inline docs/001_develop/03_client-capabilities/011_criteria/05_serialisers.mdx:18]: Same comment with this table

**Review by matteematt — COMMENTED** [review]
All looks great, just a quick tidy up with some formatting and links

Though the questions about the form criteria page defo need looking at

File: docs/001_develop/03_client-capabilities/011_criteria/04_form-criteria-builder.mdx (hunk 6 above)
Comment (patrickoneill-genesis) [reply docs/001_develop/03_client-capabilities/011_criteria/04_form-criteria-builder.mdx:14]: Not really clear on what this class does (or if it is even that useful, leaving the page off for now)

File: docs/001_develop/03_client-capabilities/011_criteria/02_criteria-builder.mdx (hunk 4 above)
Comment (patrickoneill-genesis) [reply docs/001_develop/03_client-capabilities/011_criteria/02_criteria-builder.mdx:33]: Fixed

File: docs/001_develop/03_client-capabilities/011_criteria/05_serialisers.mdx (hunk 7 above)
Comment (patrickoneill-genesis) [reply docs/001_develop/03_client-capabilities/011_criteria/05_serialisers.mdx:19]: Fixed

File: docs/001_develop/03_client-capabilities/011_criteria/index.mdx (hunk 2 above)
Comment (matteematt) [reply docs/001_develop/03_client-capabilities/011_criteria/index.mdx:42]: Did you mean to mark this as resolved as it looks like you didn't remove it yet?

**Review by matteematt — COMMENTED** [review]
Thanks for making the changes. Just one small bit to go

File: docs/001_develop/03_client-capabilities/011_criteria/index.mdx (hunk 2 above)
Comment (patrickoneill-genesis) [reply docs/001_develop/03_client-capabilities/011_criteria/index.mdx:42]: This is being auto generated from the foundation ui module. moved it to docs/index.mdx and renamed overview to index.mdx

File: docs/001_develop/03_client-capabilities/011_criteria/01_overview.mdx
Comment (patrickoneill-genesis) [reply docs/001_develop/03_client-capabilities/011_criteria/01_overview.mdx:1]: Changed to index.mdx and moved the generated one to docs/index.mdx

**Review by matteematt — APPROVED** [review]

**Pull request merged**